src/DefaultBuilder/src/WebApplicationOptions.cs

```diff
@@ -64,5 +64,39 @@ internal void ApplyHostConfiguration(IConfigurationBuilder builder)
                 builder.AddInMemoryCollection(config);
             }
         }
+
+        internal void ApplyApplicationName(IWebHostBuilder webHostBuilder)
+        {
+            string? applicationName = null;
+
+            // We need to "parse" the args here since
+            // we need to set the application name via UseSetting
+            if (Args is not null)
+            {
+                var config = new ConfigurationBuilder()
+                        .AddCommandLine(Args)
+                        .Build();
+
+                applicationName = config[WebHostDefaults.ApplicationKey];
+
+                // This isn't super important since we're not adding any disposable sources
+                // but just in case
+                if (config is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+            }
+
+            // Application name overrides args
+            if (ApplicationName is not null)
+            {
+                applicationName = ApplicationName;
+            }
+
+            if (applicationName is not null)
+            {
+                webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, applicationName);
+            }
+        }
```

Comment:
Nit: You don't have build config if ApplicationName is not null. And I think it'd be cleaner to move the `Assembly.GetEntryAssembly()?.GetName()?.Name` fallback into this method so you see the whole picture.

```suggestion
        internal void ApplyApplicationName(IWebHostBuilder webHostBuilder)
        {
            // ApplicationName overrides everything else
            string? applicationName = ApplicationName;

            // Fall back to args if ApplicationName wasn't specified
            if (applicationName is null && Args is not null)
            {
                var config = new ConfigurationBuilder()
                        .AddCommandLine(Args)
                        .Build();

                applicationName = config[WebHostDefaults.ApplicationKey];

                // This isn't super important since we're not adding any disposable sources
                // but just in case
                (config as IDisposable)?.Dispose();
            }

            // Lastly fall back to the entry Assembly name.
            applicationName ??= Assembly.GetEntryAssembly()?.GetName()?.Name ?? string.Empty;

            webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, applicationName);
        }
```

Reply:
I didn't try/test this. If you take it we should remove the call to `webHostBuilder.UseSetting(WebHostDefaults.ApplicationKey, ...)` in WebApplicationBuilder of course.

Reply:
This has args overriding the application name. I like the change overall though.